### PR TITLE
Enhancement: Add workflow for creating release and uploading release asset

### LINKED ISSUE #22: automate release asset upload
`cs2pr` is bundled with https://github.com/shivammathur/setup-php.
for this to work, https://github.com/shivammathur/setup-php needs a "cs2pr" release asset bundled with every release we make.

this asset upload is currently done manually (and therefore sometimes forgotten).

Maybe we can automate this using https://github.com/actions/upload-release-asset ?
(or with any other action?)

triggered by https://github.com/staabm/annotate-pull-request-from-checkstyle/issues/17

### PR DESCRIPTION
**Pull request by localheinz**
This PR

* [x] adds a workflow for creating a release and uploading a release asset

Fixes #22.

### CODE REVIEW

File: .github/workflows/continuous-deployment.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+      - name: "Upload cs2pr"
+        uses: "actions/upload-release-asset@v1"
+        env:
+          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
+        with:
+          asset_content_type: "text/plain"
+          asset_name: "cs2pr"
+          asset_path: "cs2pr"
+          upload_url: "${{ steps.create-release.outputs.upload_url }}"
```
Comment (localheinz) [inline .github/workflows/continuous-deployment.yml]: For reference, see https://github.com/ergebnis/composer-normalize/blob/435ba576a79bb00ae16105ef674d72430612bd7d/.github/workflows/continuous-deployment.yml - it's almost the same workflow.

**Comment by staabm** [comment]
thx for the fast PR :-).

does this mean I will only create a git tag and the release itself will be created automatically, right?

**Comment by localheinz** [comment]
> does this mean I will only create a git tag and the release itself will be created automatically, right?

👍 

If you want to use your personal access token, then it will even appear as if you released it (instead of GitHub Actions).

**Comment by localheinz** [comment]
> If you want to use your personal access token, then it will even appear as if you released it (instead of GitHub Actions).

Should be fine as is - as observed with 25fefe481c142e16c74ce6fbc1331e3e0221c335!

File: .github/workflows/continuous-deployment.yml [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        with:
+          draft: false
+          prerelease: false
+          release_name: "${{ steps.determine-tag.outputs.tag }}"
+          tag_name: "${{ steps.determine-tag.outputs.tag }}"
+
+      - name: "Upload cs2pr"
+        uses: "actions/upload-release-asset@v1"
+        env:
+          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
```
Comment (staabm) [inline .github/workflows/continuous-deployment.yml]:
```suggestion
          GITHUB_TOKEN: "${{ secrets.STAABM_TOKEN }}"
```

File: .github/workflows/continuous-deployment.yml [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+      - name: "Determine tag"
+        id: "determine-tag"
+        run: "echo \"::set-output name=tag::${GITHUB_REF#refs/tags/}\""
+
+      - name: "Create release"
+        id: "create-release"
+        uses: "actions/create-release@v1.0.0"
+        env:
+          GITHUB_TOKEN: "${{ secrets.GITHUB_TOKEN }}"
```
Comment (staabm) [inline .github/workflows/continuous-deployment.yml]:
```suggestion
          GITHUB_TOKEN: "${{ secrets.STAABM_TOKEN }}"
```

**Comment by staabm** [comment]
lets try that, thx!

**Comment by localheinz** [comment]
Thank you, @staabm!

**Comment by staabm** [comment]
Hmmm when creating a release via the github.com website the build starts and errors, see https://github.com/staabm/annotate-pull-request-from-checkstyle/runs/446847091?check_suite_focus=true

Is there a way to add the file to the already existing release in this case, instead of trying to add a new one?

Related issue https://github.com/actions/create-release/issues/41

**Comment by shivammathur** [comment]
@localheinz @staabm 
This should run on release event and not on each push, so that it does not create a new release for each push, instead pushes the asset to already existing release.

```yaml
on:
  release:
    types: [published]
```

Also remove the `create_release` action. You can get `upload_url` from the latest release using the GitHub API because the release is already created and only the asset has to be uploaded.
https://developer.github.com/v3/repos/releases/#get-the-latest-release

**Comment by shivammathur** [comment]
You can use `$GITHUB_EVENT_PATH` variable as well to get the `upload_url`.
```yaml
- name: Get upload_url
  run: echo "::set-output name=upload_url::https://uploads.github.com/repos/$GITHUB_REPOSITORY/releases/$(jq --raw-output '.release.id' $GITHUB_EVENT_PATH)/assets{?name,label}"  
  id: release

- name: "Upload cs2pr"
  uses: "actions/upload-release-asset@v1"
  env:
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
  with:
    asset_content_type: "text/plain"
    asset_name: "cs2pr"
    asset_path: "cs2pr"  
    upload_url: ${{ steps.release.outputs.upload_url }}
```

**Comment by localheinz** [comment]
@shivammathur @staabm 

This has worked fine for me as so far I have created releases by tagging locally and pushing tags to `origin`. Looks like it doesn’t when the release is created using the GitHub Ui.

**Comment by shivammathur** [comment]
ok, just noticed that it runs only on tags. My bad.

For releases created using GitHub UI the release is already created, so in workflow it does not make sense to create a new one, so for this can you can add another workflow using  https://github.com/staabm/annotate-pull-request-from-checkstyle/pull/23#issuecomment-586542973

**Comment by staabm** [comment]
My regular release-flow would not involve creating tags locally and push them. 

I am fine with adjusting the existing GithubAction workflow for the github.com UI release flow.

**Comment by staabm** [comment]
hmm it still doesnt work https://github.com/staabm/annotate-pull-request-from-checkstyle/issues/40

**Comment by shivammathur** [comment]
@staabm 

Please change the event from tags to release, so that `$GITHUB_EVENT_PATH` has release info.
```diff
- on:
-   push:
-     tags:
-       - "**"
+ on:
+   release:
+     types:
+       - published
```

**Comment by staabm** [comment]
thx @shivammathur. will try your suggestion in https://github.com/staabm/annotate-pull-request-from-checkstyle/pull/41 <3

**Pull request merged**